pulsar-client-go/pulsar/binaryfreelist.go

```diff
@@ -0,0 +1,318 @@
+		value = float64(v)
+	case int:
+		if value = float64(v); int(value) != v {
+			return nil, fmt.Errorf("serialize failed: provided Go int would lose precision: %d", v)
+		}
+	case int64:
+		if value = float64(v); int64(value) != v {
+			return nil, fmt.Errorf("serialize failed: provided Go int64 would lose precision: %d", v)
+		}
+	case int32:
+		if value = float64(v); int32(value) != v {
+			return nil, fmt.Errorf("serialize failed: provided Go int32 would lose precision: %d", v)
+		}
+	default:
+		return nil, fmt.Errorf("serialize failed: expected: Go numeric; received: %T", datum)
+	}
+	var buf []byte
+	buf = append(buf, 0, 0, 0, 0, 0, 0, 0, 0)
+	binary.BigEndian.PutUint64(buf[len(buf)-8:], math.Float64bits(value))
+	return buf, nil
+}
+
+func (b BinaryFreeList) PutFloat32(datum interface{}) ([]byte, error) {
+	var value float32
+	switch v := datum.(type) {
+	case float32:
+		value = v
+	case float64:
+		value = float32(v)
+	case int:
+		if value = float32(v); int(value) != v {
+			return nil, fmt.Errorf("serialize failed: provided Go int would lose precision: %d", v)
+		}
+	case int64:
+		if value = float32(v); int64(value) != v {
+			return nil, fmt.Errorf("serialize failed: provided Go int64 would lose precision: %d", v)
+		}
+	case int32:
+		if value = float32(v); int32(value) != v {
+			return nil, fmt.Errorf("serialize failed: provided Go int32 would lose precision: %d", v)
+		}
+	default:
+		return nil, fmt.Errorf("serialize failed: expected: Go numeric; received: %T", datum)
+	}
+	var buf []byte
+	buf = append(buf, 0, 0, 0, 0)
+	binary.BigEndian.PutUint32(buf[len(buf)-4:], uint32(math.Float32bits(value)))
+	return buf, nil
+}
+
+func ReadElements(r io.Reader, elements ...interface{}) error {
+	for _, element := range elements {
+		err := readElement(r, element)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func WriteElements(w io.Writer, elements ...interface{}) error {
+	for _, element := range elements {
+		err := writeElement(w, element)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func readElement(r io.Reader, element interface{}) error {
+	switch e := element.(type) {
+	case *int8:
+		rv, err := BinarySerializer.Uint8(r)
+		if err != nil {
+			return err
+		}
+		*e = int8(rv)
+		return nil
+
+	case *int16:
+		rv, err := BinarySerializer.Uint16(r, bigEndian)
+		if err != nil {
+			return err
+		}
+		*e = int16(rv)
+		return nil
+
+	case *int32:
+		rv, err := BinarySerializer.Uint32(r, bigEndian)
+		if err != nil {
+			return err
+		}
+		*e = int32(rv)
+		return nil
+
+	case *int64:
+		rv, err := BinarySerializer.Uint64(r, bigEndian)
+		if err != nil {
+			return err
+		}
+		*e = int64(rv)
+		return nil
+
+	case *bool:
+		rv, err := BinarySerializer.Uint8(r)
+		if err != nil {
+			return err
+		}
+		if rv == 0x00 {
+			*e = false
+		} else {
+			*e = true
+		}
+		return nil
+	}
+	return binary.Read(r, bigEndian, element)
+}
+
+func writeElement(w io.Writer, element interface{}) error {
```

Comment:
it would be good to think about how to add integration tests between java and go. basically the content serialized by java should be deseralized by go. vice versa.

Reply:
yes, good idea.